contracts/eosio.system/include/eosio.system/eosio.system.hpp

```diff
@@ -1172,6 +1220,66 @@ namespace eosiosystem {
          [[eosio::action]]
          void unregprod( const name& producer );
 
+         /**
+          * Action to permanently transition to Savanna consensus.
+          * Create the first generation of finalizer policy and activate
+          * the policy by using `set_finalizers` host function 
+          *
+          * @pre Require the authority of the contract itself
+          * @pre A sufficient number (15) of the top 21 block producers have registered a finalizer key
+          */
+         [[eosio::action]]
+         void switchtosvnn();
+
+         /**
+          * Register a finalizer key action by a registered producer.
+          * If this is the first registered finalizer key of the producer,
+          * it will also implicitly be marked active.
+          * If this finalizer key was registered before (and still exists) even
+          * by other block producers, the registration will fail.
+          * A registered producer can have multiple registered finalizer keys.
+          *
+          * @param producer - account registering `finalizer_key`,
+          * @param finalizer_key - key to be registered. The key is in base64url format.
+          * @param proof_of_possession - a valid Proof of Possession signature to show the producer owns the private key of the finalizer_key. The signature is in base64url format.
+          *
+          * @pre `producer` must be a registered producer
+          * @pre Authority of `producer` to register
+          */
+         [[eosio::action]]
+         void regfinkey( const name& finalizer, const std::string& finalizer_key, const std::string& proof_of_possession);
+
+         /**
+          * Activate a finalizer key action. If the block producer is currently
+          * active (in top 21), then immediately change Finalizer Policy.
+          * Activating a finalizer key of a block producer implicitly deactivates
+          * the previously active finalizer key of that block producer.
+          *
+          * @param producer - account activating `finalizer_key`,
+          * @param finalizer_key - key to be activated.
+          *
+          * @pre `producer` must be a registered producer
+          * @pre `finalizer_key` must be a registered  finalizer key
+          * @pre Authority of `producer`
```

Comment:
Should we note something about using linkauth to allow this by a lower authority?